### DISCUSSION TITLE
How do I close a TCP connection?

### DISCUSSION THREAD

**Discussion by tilongzs** [cesanta/mongoose#1638] [Ask questions, get answers] [answered]
- My goal is: The server gets the message that the client is disconnected from the server.
- My actions were: mg_close_conn(_currentConn);
- My expectation was: The client executes mg_close_conn(conn); mg_mgr_free(mgr); code, and the server receives the MG_EV_CLOSE event.
- The result I saw: The client executes mg_close_conn(conn); mg_mgr_free(mgr); code, the client receives the MG_EV_CLOSE event, but the server does not receive the MG_EV_CLOSE event. The server side receives the MG_EV_CLOSE event only when the client program is completely closed.
- My question is: Which step am I missing?

# Environment

- mongoose version: 7.7
- Compiler/IDE and SDK: Visual Studio 2022 (v143)  Windows SDK 10.0.19041.0
- Target hardware/board: Windows 11 X64(10.0.22000.795)

**Comment by scaprile** [comment]
Please correct me where my guessing was off.
You have a server and a client on a TCP port. Both implemented using Mongoose, and hopefully running on different devices <sup>1</sup>.
Mongoose function calls mostly "request" an action to be done, the action itself gets actually done by the event manager, particularly on TCP where there are handshakes to be performed. So, you can't stop calling the manager and free its resources until you are sure your job is done.
<sup>1</sup> (If both client and server are on the same device and your client exits, your server might not have processed the event when you exit the event manager)

A proper client disconnect can be observed on the [HTTP client example](https://github.com/cesanta/mongoose/tree/master/examples/http-client) and is explained in detail in the [error handling tutorial](https://mongoose.ws/tutorials/error-handling/https://mongoose.ws/tutorials/error-handling/)

**Comment by tilongzs** [comment]
> Please correct me where my guessing was off. You have a server and a client on a TCP port. Both implemented using Mongoose, and hopefully running on different devices 1. Mongoose function calls mostly "request" an action to be done, the action itself gets actually done by the event manager, particularly on TCP where there are handshakes to be performed. So, you can't stop calling the manager and free its resources until you are sure your job is done. 1 (If both client and server are on the same device and your client exits, your server might not have processed the event when you exit the event manager)
> 
> A proper client disconnect can be observed on the [HTTP client example](https://github.com/cesanta/mongoose/tree/master/examples/http-client) and is explained in detail in the [error handling tutorial](https://mongoose.ws/tutorials/error-handling/https://mongoose.ws/tutorials/error-handling/)

I re-ran the test, both sides are using Mongoose, when the server and client are running on different devices, the server can receive the MG_EV_CLOSE event. When running on the same device, the server cannot receive the MG_EV_CLOSE event.
When running on the same device: the server side uses Mongoose and the client side does not use Mongoose, the server side can receive MG_EV_CLOSE event; when the server side does not use Mongoose and the client side uses Mongoose, the server side cannot receive the connection disconnect message.
My requirement is simple, how can the client completely disconnect TCP connection without close the client program on the same device (as said before, I have executed mg_close_conn(conn); mg_mgr_free(mgr)).

**Comment by scaprile** [comment, marked as answer]
Please check the examples and tutorials, you can not call `mg_close_conn(conn); mg_mgr_free(mgr));` like that (no example does that), as already explained.
If you don't want to exit the program, then just don't exit the main loop (and of course don't free resources). All examples that don't exit do just that.
Please post the smallest piece of code showing what you want to do and can't do.

**Reply by tilongzs** [reply]
// post the smallest piece of code 

```
// Connect the server
mg_mgr* mgr = new mg_mgr;
	mg_mgr_init(mgr);
	thread([&, mgr]
		{
			while (true)
			{
				mg_mgr_poll(mgr, 1);
			}

			mg_mgr_free(mgr);
		}).detach();

conn = mg_connect(mgr, "tcp://127.0.0.1:23300", OnTCPEvent, this);
```

// Try to close the connection a few seconds after the connection is successful
`mg_close_conn(conn);`

 The server side cannot receive the connection disconnect message.

**Reply by scaprile** [reply]
From [the docs](https://mongoose.ws/documentation/#2-minute-integration-guide)
> NOTE: Since Mongoose's core is not protected against concurrent accesses, make sure that all mg_* API functions are called from the same thread or RTOS task.

If you absolutely need multi-threading, take the time to go through the user guide (linked above) and [tutorials](https://mongoose.ws/tutorials/), there is one that shows you how to do [multi-threading](https://mongoose.ws/tutorials/multi-threaded/); though I suggest you start with something simpler.

No example calls `mg_close_conn()`, sorry about not pointing that out clearly, I edited my former response for clarity. From [the docs](https://mongoose.ws/documentation/#best-practices):
> - If you need to close the connection, set `c->is_draining = 1;` in your event handler function. That tells the event manager to send all remaining data in a send buffer ("drain" the connection), then close the connection. If you need to close the connection immediately without draining, use `c->is_closing = 1;`

[Minimal (but verbose) example showing what you want to achieve](https://github.com/cesanta/mongoose/files/9200509/tcpcs.zip)

**Reply by tilongzs** [reply]
It's done, thanks.
